libs/vkd3d/command.c

```diff
@@ -2111,6 +2111,66 @@ static void d3d12_command_list_clear_attachment_deferred(struct d3d12_command_li
         attachment->value.depthStencil.stencil = clear_value->depthStencil.stencil;
 }
 
+static void d3d12_command_list_discard_attachment_barrier(struct d3d12_command_list *list,
+        struct d3d12_resource *resource, const VkImageSubresourceLayers *subresource, bool is_bound)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    const struct vkd3d_format *format;
+    VkImageMemoryBarrier barrier;
+    VkPipelineStageFlags stages;
+    VkAccessFlags access;
+    VkImageLayout layout;
+
+    if (resource->desc.Flags & D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET)
+    {
+        stages = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
+        access = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+        layout = is_bound ? d3d12_resource_pick_layout(resource, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL) : resource->common_layout;
+    }
+    else if (resource->desc.Flags & D3D12_RESOURCE_FLAG_ALLOW_DEPTH_STENCIL)
+    {
+        stages = VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT;
+        access = VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
+        layout = is_bound && list->dsv_layout ? list->dsv_layout : resource->common_layout;
+    }
+    else
+    {
+        ERR("Unsupported resource flags %#x.\n", resource->desc.Flags);
+        return;
+    }
+
+    /* Transitioning only one aspect of a depth-stencil image is not allowed */
+    format = vkd3d_format_from_d3d12_resource_desc(list->device, &resource->desc, 0);
+
+    if (format->vk_aspect_mask != subresource->aspectMask)
+        return;
+
+    barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+    barrier.pNext = NULL;
+    barrier.srcAccessMask = 0;
```

Comment:
If we're in a suspended render pass, is it safe to use 0 here for srcAccessMask?

Reply:
We're discarding old content, so we have no reason to flush it.

Reply:
Technically, memory needs to be available to be able to safely do a layout transition, even from UNDEFINED. I guess the possible hazard would be that this chain of events could get mangled:

- Write something (A)
- Discard
- Write something (B)

The problem scenario is if there are some pending writes in A, which are only resolved after B writes resolve.

Reply:
This seems a bit pedantic though. I'll be okay doing 0 here if it's commented why it's 0.